crates/macro/src/html_tree/html_component.rs

```diff
@@ -360,181 +364,97 @@ impl Props {
-        Ok(props_type)
-    }
-}
-
-impl PeekValue<PropType> for Props {
-    fn peek(cursor: Cursor) -> Option<PropType> {
-        let (ident, _) = cursor.ident()?;
 
-        let prop_type = if ident == "with" {
-            PropType::With
-        } else {
-            PropType::List
-        };
-
-        Some(prop_type)
-    }
-}
-
-impl Parse for Props {
-    fn parse(input: ParseStream) -> ParseResult<Self> {
-        match Props::get_type(input)? {
-            Some(PropType::List) => input.parse().map(|l| Props::List(Box::new(l))),
-            Some(PropType::With) => {
-                let result = input.parse().map(|w| Props::With(Box::new(w)));
-                match result {
-                    Ok(v) => Ok(v),
-                    Err(e) => {
-                        println!("MY ERROR");
-                        return Err(input.error("MY"));
-                    }
-                }
+            if (HtmlProp::peek(input.cursor())).is_none() {
+                break;
             }
-            None => Ok(Props::None),
-        }
-    }
-}
 
-struct ListProps {
-    props: Vec<HtmlProp>,
-    node_ref: Option<Expr>,
-}
+            let prop = input.parse::<HtmlProp>()?;
+            if prop.label.to_string() == "ref" {
+                match node_ref {
+                    None => Ok(()),
+                    Some(_) => Err(syn::Error::new_spanned(&prop.label, "too many refs set")),
+                }?;
 
-impl Parse for ListProps {
-    fn parse(input: ParseStream) -> ParseResult<Self> {
-        let ListProps {
-            mut props,
-            node_ref,
-        } = Props::remove_refs(Props::collect_props(input)?);
-
-        // alphabetize
-        props.sort_by(|a, b| {
-            if a.label == b.label {
-                Ordering::Equal
-            } else if a.label.to_string() == "children" {
-                Ordering::Greater
-            } else if b.label.to_string() == "children" {
-                Ordering::Less
-            } else {
-                a.label
-                    .to_string()
-                    .partial_cmp(&b.label.to_string())
-                    .unwrap()
+                node_ref = Some(prop.value);
+                continue;
             }
-        });
 
-        Ok(ListProps { props, node_ref })
-    }
-}
+            if prop.label.to_string() == "type" {
+                return Err(syn::Error::new_spanned(&prop.label, "expected identifier"));
+            }
 
-struct WithProps {
-    props: Ident,
-    node_ref: Option<Expr>,
-}
+            if !prop.label.extended.is_empty() {
+                return Err(syn::Error::new_spanned(&prop.label, "expected identifier"));
+            }
 
-impl Parse for WithProps {
-    fn parse(input: ParseStream) -> ParseResult<Self> {
-        let with = input.parse::<Ident>()?;
-        if with != "with" {
-            return Err(input.error("expected to find `with` token"));
+            match props {
+                ref mut props @ Props::None => {
```

Comment:
@jstarry Coul you please explain what does it mean `@` ?

Reply:
Check this out: https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#-bindings

It allows you to reference the original pattern-matched value. In this case, I use it to set `props` in a match branch :)

Reply:
Thanks, I'm only Rust beginner. I'm very appreciate your help

Reply:
Happy to help! I also learned Rust by helping with this project 😄 Thanks for your help!